powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/SignatureServiceBehavior.java

```diff
@@ -390,6 +418,19 @@ private void handleValidSignature(ActivationRecordEntity activation, ValidateSig
         // Get ActivationRepository
         final ActivationRepository activationRepository = repositoryCatalogue.getActivationRepository();
 
+        if (activation.getVersion() == 3) {
+            HashBasedCounter hashBasedCounter = new HashBasedCounter();
+            // Number of ctr_data increments is the lowest counter + 1 - current counter
+            // (+1, since the client has incremented the counter)
+            long incrementCount = validationResponse.getLowestValidCounter() + 1 - activation.getCounter();
+            byte[] ctrData = BaseEncoding.base64().decode(activation.getCtrDataBase64());
```

Comment:
I think we should pass the `ctrData` here, just as we do with counter (we use it only for v3) - otherwise, we iterate the hash twice, which is not necessary.